swift/Arrow/Sources/Arrow/ArrowCImporter.swift

```diff
@@ -0,0 +1,143 @@
+        pointer.copyMemory(from: cBuffer!, byteCount: byteCount)
+        arrowBuffers.append(ArrowBuffer(length: length, capacity: UInt(byteCount), rawPointer: pointer))
+    }
+
+    public init() {}
+
+    public func importType(_ cArrow: String, name: String = "") ->
+        Result<ArrowField, ArrowError> {
+        do {
+            let type = try ArrowType.fromCDataFormatId(cArrow)
+            return .success(ArrowField(name, type: ArrowType(type.info), isNullable: true))
+        } catch {
+            return .failure(.invalid("\(error)"))
+        }
+    }
+
+    public func importField(_ cSchema: ArrowC.ArrowSchema) ->
+        Result<ArrowField, ArrowError> {
+        if cSchema.n_children > 0 {
+            return .failure(.invalid("Children currently not supported"))
+        } else if cSchema.dictionary != nil {
+            return .failure(.invalid("Dictinoary types currently not supported"))
+        }
+
+        switch importType(
+            String(cString: cSchema.format), name: String(cString: cSchema.name)) {
+        case .success(let field):
+            release(cSchema)
+            return .success(field)
+        case .failure(let err):
+            return .failure(err)
+        }
+    }
+
+    public func importArray(
+        _ cArray: ArrowC.ArrowArray,
+        arrowType: ArrowType,
+        isNullable: Bool = false) -> Result<ArrowArrayHolder, ArrowError> {
+        let arrowField = ArrowField("", type: arrowType, isNullable: isNullable)
+        return importArray(cArray, arrowField: arrowField)
+    }
+
+    public func importArray(
+        _ cArray: ArrowC.ArrowArray,
+        arrowField: ArrowField) -> Result<ArrowArrayHolder, ArrowError> {
+        if cArray.null_count < 0 {
+            return .failure(.invalid("Uncomputed null count is not supported"))
+        } else if cArray.n_children > 0 {
+            return .failure(.invalid("Children currently not supported"))
+        } else if cArray.dictionary != nil {
+            return .failure(.invalid("Dictinoary types currently not supported"))
+        }
+
+        let arrowType = arrowField.type
+        let length = UInt(cArray.length)
+        let nullCount = UInt(cArray.null_count)
+        let nullBytes = Int(ceil(Double(length) / 8))
+        var arrowBuffers = [ArrowBuffer]()
+
+        if cArray.n_buffers > 0 {
+            if cArray.buffers == nil {
+                return .failure(.invalid("C array buffers is nil"))
+            }
+
+            switch arrowType.info {
+            case .variableInfo:
+                if cArray.n_buffers != 3 {
+                    return .failure(
+                        .invalid("Variable buffer count expected 3 but found \(cArray.n_buffers)"))
+                }
+
+                appendToBuffer(cArray.buffers[0], arrowBuffers: &arrowBuffers, byteCount: nullBytes, length: length,
+                               nullCount: nullCount)
+                let byteCount = MemoryLayout<Int32>.stride * Int(length)
```

Comment:
I don't know how far you are in implementing Arrow datatypes, but note that large types such as LargeString have 64-bit offsets.

Reply:
Also, the number of offsets is `length + 1`.

Reply:
Good point! LargeStrings are currently not supported